mesonbuild/dependencies/base.py

```diff
@@ -308,6 +312,8 @@ def __init__(self, kwargs):
     def get_for_machine_from_kwargs(self, kwargs):
         return MachineChoice.BUILD if kwargs.get('native', False) else MachineChoice.HOST
 
+msystem = detect_msystem()
```

Comment:
I think it probably makes more sense to add an `is_msys` to the `MachineInfo` class in `envconfig.py` and possibly one in `mesonlib.py` as well. then in the dependency object itself you can do something like:

```python
if self.environment.machines[self.for_machine].is_msys():
    ....
````

Reply:
You may also want and `is_mingw` method, after reading through this more

Reply:
What does for_machine means? does for_machine means the target platform? for this, we are talking about native platform, not the target platform, that's different.

Reply:
`for_machine` is usually a `mesonlib.MachineChoice`, which means that it tells whether this compiler/dependency targets the build machine or the host machine. You could, for example, be using an msys compiler as your "build" compiler and msvc as your "host" compiler, for example when building for arm based windows.

Reply:
This is the only thing left I'm not sure about, everything else looks good. @Ericson2314, do you have an opinion on whether this is fine or we should add it to the `MachineInfo` class?

Reply:
> This is the only thing left I'm not sure about, everything else looks good. @Ericson2314, do you have an opinion on whether this is fine or we should add it to the `MachineInfo` class?

I still don't know how to add it to the MachineInfo class. Because I am retrieving host info, not the guest info

Reply:
ping

Reply:
`__CYGWIN__` is defined in both msys2/cygwin, `__MSYS__` is defined only in msys2.

Reply:
> I did a bit of googling, you can use the compiler to determine whether you're using a mingw compiler, because `__MINGW32__` will bedefined. I'm not sure about msys. But the more I think about it, the more I think we need to handl this in the MachineInfo class, because if we don't then this _only_ works for the build machine.
> 
> @lazka, is there an easy way for us to figure out if we're running in an msys2 environment from the compiler? or some other way than just "is an env var defined"? I want to be able to make sure that we're associating the correct machine (the host or the build) with msys.

Gotcha, using the compiler to detecting __MINGW32__ is a good idea, I may use that, but din't know where to start yet.
So if the platform is windows, and dtecting the compiler macro, then we using cygpath to converting paths?
Or we could implement the path convert directly using python script?

Reply:
The cygpath belongs to msys2 runtime, so if msys2 present, it always present. unless we doesn't want to using msys2 runtime at all
https://packages.msys2.org/package/msys2-runtime?repo=msys&variant=x86_64

Reply:
> `__CYGWIN__` is defined in both msys2/cygwin, `__MSYS__` is defined only in msys2.

So all of these can be deteced by using the compiler

Reply:
I'm not a fan of hardcoding msys2 specifics in meson, but other than that: if windows and the path starts with "/mingw" and cygpath is in PATH would be a good indicator.

Reply:
> I'm not a fan of hardcoding msys2 specifics in meson, but other than that: if windows and the path starts with "/mingw" and cygpath is in PATH would be a good indicator for converting paths. But not 100% sure what you are trying to do here.

I am trying to converting path from /mingw32/include/a to windows' style absolute path, sometimes the mingw compiler doesn't recognize the path like -I/mingw32/include/a

Reply:
> I'm not a fan of hardcoding msys2 specifics in meson, but other than that: if windows and the path starts with "/mingw" and cygpath is in PATH would be a good indicator for converting paths (or maybe the path not existing, but the converted path existing). But not 100% sure what you are trying to do here.

by the way, if cygpath are already present, then the is no need convert it by myself.

Reply:
Hi, I am confusing, does MachineInfo system standfor? are the system stand for the target system or the host system?

def detect_machine_info(env: T.Optional['Environment'] = None, compilers: T.Optional[CompilersDict] = None) -> MachineInfo:
    """Detect the machine we're running on

    If compilers are not provided, we cannot know as much. None out those
    fields to avoid accidentally depending on partial knowledge. The
    underlying ''detect_*'' method can be called to explicitly use the
    partial information.
    """
    if env and compilers:
        print('Start print')
        for compiler in compilers.values():
            v, _ = compiler.get_define('__MINGW32__',  '', env, [], [])
            print(compiler)
            print(v)

    return MachineInfo(
        detect_system(),
        detect_cpu_family(compilers) if compilers is not None else None,
        detect_cpu(compilers) if compilers is not None else None,
        sys.byteorder)

I can using compiler to retrive the __MINGW32__ macro info

Reply:
How to setup the native build for detect_system()

Reply:
ping for suggestion. there is no system such as mingw and msys, how to add it and dosn't affect things?

Reply:
> Hi, I am confusing, does MachineInfo system standfor? are the system stand for the target system or the host system?

@lygstate we have only tried to auto-detect assuming native compilation, but I am not too pleased with the way auto-detection works today (guess platform, find compiler, guess platform harder using compiler). If you have some other ideas I'd certainly be open to it

Reply:
@dcbaker I think I would prefer it go in `MachineInfo`. I wish we didn't have any top-level `is_*` functions either, but I guess I never removed those.

Reply:
> @dcbaker I think I would prefer it go in `MachineInfo`. I wish we didn't have any top-level `is_*` functions either, but I guess I never removed those.

no problem, but how? maybe we need a more parameter such as vendor in MachineInfo? cygwin/msys2 almost the same but different vendor; msvc/intel-for-msvc/clang-for-msvc almost the same but different vendor;  msys2/mingw are also a compier for windows? how to distinguish them properly in machineInfo?

Reply:
ping for help

Reply:
If what @lazka says above is true and this issue is fixed in msys2, how much of this MR is still needed?